File: commons-mis/src/main/java/io/github/cooperlyt/mis/work/WorkRemoteService.java
package io.github.cooperlyt.mis.work;


import io.github.cooperlyt.mis.work.data.WorkDefine;
import io.github.cooperlyt.mis.work.data.WorkDefineForCreate;
import io.github.cooperlyt.mis.work.data.WorkDefineForProcess;
import reactor.core.publisher.Mono;

import java.util.Map;

public interface WorkRemoteService {


  Mono<WorkDefineForCreate> prepareCreate(String defineId);

  Mono<WorkDefineForCreate> recreate(String defineId, long originalWorkId);

  Mono<WorkDefineForProcess> prepareProcess(String defineId);

  Mono<WorkDefine> define(String defineId);

//  @Transactional
//  Mono<Long> runWork(String bindingName, String defineId, long orgId, long workId, Function<WorkDefine, Mono<Map<String,Object>>> dataProcess);
//
////  @Transactional
////  <T> Mono<T> createWork(String defineId, String bindingName, Function<WorkCreateProvide, Mono<T>> dataProcess);
//
//  @Transactional
//  <T> Mono<T> createWork(String bindingName, String defineId, long orgId,
//                         WorkCreateType type, Function<WorkDefineForCreate, Mono<T>> dataProcess);
//
//
//  Mono<Void> createOptionalWork(String bindingName, String defineId, long orgId,
//                                WorkCreateType type, Function<WorkDefineForCreate, Mono<Boolean>> dataProcess);
//
//
//  Mono<Long> sendWorkMessage(String bindingName, String defineId, long workId);


  Mono<Long> sendWorkMessage(String bindingName, String defineId,
                             long workId, Map<String,Object> processData);

  Mono<Long> applyWorkId();

}

File: commons-mis/src/main/java/io/github/cooperlyt/mis/work/WorkRemoteService.kt
package io.github.cooperlyt.mis.work

import io.github.cooperlyt.mis.work.data.WorkDefine
import io.github.cooperlyt.mis.work.data.WorkDefineForCreate
import io.github.cooperlyt.mis.work.data.WorkDefineForProcess
import io.github.cooperlyt.mis.work.message.ProcessDocumentation
import io.github.cooperlyt.mis.work.message.WorkCreateMessage
import io.github.cooperlyt.mis.work.message.WorkEventMessage
import org.springframework.messaging.Message
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.util.function.Supplier

interface WorkRemoteService {

    fun prepareCreate(defineId: String): Mono<WorkDefineForCreate>

    fun recreate(defineId: String, originalWorkId: Long): Mono<WorkDefineForCreate>

    fun prepareProcess(defineId: String): Mono<WorkDefineForProcess>

    fun define(defineId: String): Mono<WorkDefine>

    fun sendWorkCreateMessage(
        defineId: String,
        workId: Long,
        processData: Map<String, Any>,
        documentation: ProcessDocumentation? = null,
    ): Mono<Long>


    fun sendWorkEventMessage(
        messageName: String,
        defineId: String,
        workId: Long,
        processData: Map<String, Any>
    ): Mono<Long>

    fun workCreateMessageSinks(): Supplier<Flux<Message<WorkCreateMessage>>>

    fun workEventMessageSinks(): Supplier<Flux<Message<WorkEventMessage>>>

    fun applyWorkId(): Mono<Long>
}